Model aware routing

jkrauss82 opened the pull request:
Related issue and conversation: https://github.com/distantmagic/paddler/issues/60

Closes https://github.com/distantmagic/paddler/issues/60

Commit:
wip on adding model routing

Commit:
revert unnecessary change

Commit:
avoid consuming the request body for upstream

jkrauss82 committed:
Merge branch 'distantmagic:main' into model-aware-routing

Commit:
sync with main

Commit:
add responses for unsupported/missing model parameter, add regex check to handle only first 64k characters available in buffer

Commit:
sync with main

Commit:
Merge branch 'main' into model-aware-routing

Commit:
fix mock_status_update.rs

Commit:
Merge branch 'main' into model-aware-routing and add model column to web dashboard

Commit:
Merge branch 'main' into model-aware-routing

jkrauss82 committed:
Merge branch 'distantmagic:main' into model-aware-routing

Commit:
fix returning model not supported on all slots for model currently blocked

Commit:
Merge branch 'main' into model-aware-routing

Commit:
make model nullable for dashboard

mcharytoniuk commented:
@jkrauss82 Thanks a lot!

1. Purely because we want to make the supervisor an optional feature.
2. I don't get this one entirely. How would an integrated cluster role be different than what we have now/planned?

jkrauss82 commented:
> I don't get this one entirely. How would an integrated cluster role be different than what we have now/planned?

If I understand the planned architecture correctly, it has a cluster service running before one or more fleets. The cluster service would take care of routing incoming requests to the appropriate fleet where then a balancer assigns a slot.

If the balancer role would have knowledge of all fleets (or maybe even all llamacpp instances) in the cluster, it could integrate the routing part as well.

So looking at it this way, my question is more about the fleet concept which adds another layer to the architecture, because without fleets the cluster service would probably not be required. But my use case is for a comparably small number of llamacpp instances and requests and I can imagine that for larger setups having the fleet concept will be very welcome.

mcharytoniuk commented:
@jkrauss82 Yes, we planned fleets and clusters around scalable deployments, but in general, I think your use case makes sense, and I've been thinking for some time about how to approach this, because I think both features are useful (fleets and some kind of routing based on parameters).

What I had an issue with was that it seemed out of place to me to add parameter-based routing to a balancer (I might be biased on that, though).

I just realized that what you are trying to do is more like a gateway than a balancer (directing traffic to specific hosts based on a specific parameter), which we can add. I have an idea on how we can implement this, so fleets stay as they are, and we can also handle your use case.

I think we need to add a feature, `paddler gateway`, that will handle such traffic. We can reuse agents to connect to it instead of a balancer, so we would have something like this:

```mermaid
flowchart 
 gateway("Gateway")
 agent_1("Agent 1")
 agent_2("Agent 2")
 llama_1("Llama-server 1: Mistral")
 llama_2("Llama-server 2: Qwen")
 agent_1 --> gateway
 agent_2 --> gateway
 llama_1 --> agent_1
 llama_2 --> agent_2
```

So, in general, that would mean Paddler could operate in two modes:
- `balancer` mode (as it is now), to just direct requests to whichever llama instance has the most available slots
- `gateway` mode (to direct traffic based on the requested model); it can also potentially integrate with a balancer
Setting up a gateway should be as simple as setting up the balancer, but it could have an entirely different routing algorithm and behavior (it can still have some elements of balancing requests between hosts with identical models).

So I would be happy to do it that way, this we will be able to continue working on fleets, and scalable deployments (with the assumption that the entire fleet will have the same model loaded in every llama-server instance), and at the same time we can have another feature (a Gateway) that can handle multiple models the way you described (maybe fallback to 3rd party APIs even?) as an additional feature.

So, if you are open to it, I would do it this way. :) That would pretty much require adding an additional command line action `paddler gateway`, moving your code to a gateway proxy, and reusing some code from the balancer. I will be happy to work with you on that and help along the way as much as necessary.

What do you think? :)

Commit:
add handling of non-utf8 chunk in first 64k bytes of request body

Commit:
Merge branch 'main' into model-aware-routing

rzerres commented:
Hey @mcharytoniuk , hey @jkrauss82 

Since a couple of weeks have passed, any new update towards the `Fleed` development or  a rebase of the given PR?
I do work on a incus instance, that has access to 2 AMD Radeon AI PRO R9700 (each 32GB VRAM). And it would be keen to make `paddler` work in a multi modal szenario.
I did build an ArchLinux package. One I'm satisfied I will upload it to AUR.

Commit:
add sort option to balancer

Commit:
fix test

Commit:
fix sort not working properly

jkrauss82 committed:
Sort option in balancer

Commit:
fix: improve model detection robustness in proxy service

- Replace .unwrap().unwrap() with proper error handling to prevent
  panics on empty bodies or read errors (returns 400 instead)
- Add retry_buffer_truncated() check to reject >64KB bodies with
  a clear error message instead of silent failure
- Consolidate model extraction into single extract_model_from_body()
  function, deduplicating UTF-8 and non-UTF-8 code paths
- Fix regex to match only JSON double quotes (not single quotes)
- Compile regex once via LazyLock instead of per-request
- Initialize requested_model to None instead of Some("")
- Use idiomatic .is_none() and !bool instead of == None / == false

Commit:
fix: set idle timeout on upstream connections to prevent stale connections

Set HttpPeer idle_timeout to 30s to prevent pooled upstream connections
from accumulating indefinitely (default was None, meaning no expiry).

Commit:
chore: restructure tests directory and add model detection test suite

- Move integration_tests/ into new tests/ directory
  (tests/integration_tests/)
- Add tests/scripts/test-model-detection.sh with 19 tests covering:
  valid requests, wrong/missing models, empty/invalid bodies,
  non-JSON content types, >64KB payloads (needle-in-haystack),
  vision requests, streaming, legacy endpoints, and edge cases
- Update all references in Cargo.toml, Makefile, test-balancer.sh,
  and jarmuz/worker-prettier.mjs
- Add 'make shell_tests' target to Makefile

Commit:
docs: slim down AGENTS.md to navigation guide

Reduce from 3.5KB to 1.5KB. Keep key files map, slot tracking
semantics, test paths, and make targets. Remove verbose explanations
that LLMs can look up in source directly.

Commit:
fix: update binary paths in integration tests after directory move

Change ../target/debug/paddler to ../../target/debug/paddler in
integration test step definitions, since integration_tests/ was moved
from repo root into tests/.

Commit:
fix: correct expected slots_idle in export_metrics integration test

Test expected slots_idle=1 but actual average was 2. Pre-existing
incorrect expectation unrelated to our changes.

Commit:
fix

jkrauss82 committed:
Merge pull request #4 from jkrauss82/fix/model-detection-robustness

fix: improve model detection robustness in proxy service
